File: src/org/genyris/math/AbstractMathFunction.java
// Copyright 2008 Peter William Birch <birchb@genyis.org>
//
// This software may be used and distributed according to the terms
// of the Genyris License, in the file "LICENSE", incorporated herein by reference.
//
package org.genyris.math;

import org.genyris.core.Exp;
import org.genyris.exception.GenyrisException;
import org.genyris.interp.ApplicableFunction;
import org.genyris.interp.Closure;
import org.genyris.interp.Environment;
import org.genyris.interp.Interpreter;
import org.genyris.interp.UnboundException;

public abstract class AbstractMathFunction extends ApplicableFunction {

    private int _minExpectedNumberOfArguments;

    public AbstractMathFunction(Interpreter interp,
            String name, int minExpectedNumberOfArguments) {
        super(interp, name, true);
        _minExpectedNumberOfArguments = minExpectedNumberOfArguments;
    }

    public Exp bindAndExecute(Closure proc, Exp[] arguments,
            Environment envForBindOperations) throws GenyrisException {
        if (arguments.length < _minExpectedNumberOfArguments)
            throw new GenyrisException("Too few arguments to " + getName());
        try {
            if(arguments.length == 1) {
                return mathOperation(arguments[0]);
            }
            Exp result = arguments[0];
            for (int i = 1; i < arguments.length; i++) {
                result = mathOperation(result, arguments[i]);
            }
            return result;
        } catch (RuntimeException e) {
            throw new GenyrisException(e.getMessage());
        }
    }

    protected Exp mathOperation(Exp unary) throws GenyrisException {
        throw new GenyrisException("Bad call to mathOperation(Exp unary)");
    }

    protected Exp mathOperation(Exp a, Exp b) throws GenyrisException{
        throw new GenyrisException("Bad call to mathOperation(Exp a, Exp b)");
    }

    public static void bindFunctionsAndMethods(Interpreter interpreter) throws UnboundException, GenyrisException {
        interpreter.bindGlobalProcedureInstance(new PlusFunction(interpreter));
        interpreter.bindGlobalProcedureInstance(new MinusFunction(interpreter));
        interpreter.bindGlobalProcedureInstance(new MultiplyFunction(interpreter));
        interpreter.bindGlobalProcedureInstance(new DivideFunction(interpreter));
        interpreter.bindGlobalProcedureInstance(new RemainderFunction(interpreter));
        interpreter.bindGlobalProcedureInstance(new GreaterThanFunction(interpreter));
        interpreter.bindGlobalProcedureInstance(new LessThanFunction(interpreter));
        interpreter.bindGlobalProcedureInstance(new PowerFunction(interpreter));

    }

}

File: src/org/genyris/math/AbstractMathFunction.kt
// Copyright 2008 Peter William Birch <birchb@genyis.org>
//
// This software may be used and distributed according to the terms
// of the Genyris License, in the file "LICENSE", incorporated herein by reference.
//
package org.genyris.math

import org.genyris.core.Exp
import org.genyris.exception.GenyrisException
import org.genyris.interp.*

abstract class AbstractMathFunction(
    interp: Interpreter,
    name: String?, private val _minExpectedNumberOfArguments: Int
) : ApplicableFunction(interp, name, true) {
    @kotlin.Throws(GenyrisException::class)
    override fun bindAndExecute(
        proc: Closure?, arguments: Array<Exp?>,
        envForBindOperations: Environment?
    ): Exp? {
        if (arguments.size < _minExpectedNumberOfArguments) throw GenyrisException("Too few arguments to " + getName())
        try {
            if (arguments.size == 1) {
                return mathOperation(arguments[0])
            }
            var result = arguments[0]
            for (i in 1..<arguments.size) {
                result = mathOperation(result, arguments[i])
            }
            return result
        } catch (e: RuntimeException) {
            throw GenyrisException(e.getMessage())
        }
    }

    @kotlin.Throws(GenyrisException::class)
    protected open fun mathOperation(unary: Exp?): Exp? {
        throw GenyrisException("Bad call to mathOperation(Exp unary)")
    }

    @kotlin.Throws(GenyrisException::class)
    protected open fun mathOperation(a: Exp?, b: Exp?): Exp? {
        throw GenyrisException("Bad call to mathOperation(Exp a, Exp b)")
    }

    companion object {
        @kotlin.Throws(UnboundException::class, GenyrisException::class)
        fun bindFunctionsAndMethods(interpreter: Interpreter) {
            interpreter.bindGlobalProcedureInstance(DivideFunction(interpreter))
            interpreter.bindGlobalProcedureInstance(GreaterThanFunction(interpreter))
            interpreter.bindGlobalProcedureInstance(LessThanFunction(interpreter))
            interpreter.bindGlobalProcedureInstance(MinusFunction(interpreter))
            interpreter.bindGlobalProcedureInstance(MultiplyFunction(interpreter))
            interpreter.bindGlobalProcedureInstance(PlusFunction(interpreter))
            interpreter.bindGlobalProcedureInstance(PowerFunction(interpreter))
            interpreter.bindGlobalProcedureInstance(RemainderFunction(interpreter))
            interpreter.bindGlobalProcedureInstance(ScaleFunction(interpreter))
            interpreter.bindGlobalProcedureInstance(SquareRootFunction(interpreter))
            interpreter.bindGlobalProcedureInstance(SinFunction(interpreter))
            interpreter.bindGlobalProcedureInstance(CosFunction(interpreter))
            interpreter.bindGlobalProcedureInstance(Atan2Function(interpreter))
        }
    }
}